### PR TITLE
Validate multiple instances

### PR DESCRIPTION
**Pull request by jesper-friis**
# Description
Update `dlite-validate` to validate sources with multiple instances, without having to provide the `--id` option. 

This PR build on top of PR #985.

**Update**: Added new Python interface to the jstore implemented in C. Right now this increases the code base, but this change makes it possible to e.g. reduce `utils.instance_from_dict()` and `Instance.asdict()` to a few lines of code. An added benefit would be consistency with JSON parsing (since the json parser in C is used under the hood).

Suggests to take the code removal/replacement in a separate PR to have little more control over changes.

## Type of change
- [x] Bug fix & code cleanup
- [x] New feature
- [ ] Documentation update
- [ ] Test update

## Checklist for the reviewer
This checklist should be used as a help for the reviewer.

- [ ] Is the change limited to one issue?
- [ ] Does this PR close the issue?
- [ ] Is the code easy to read and understand?
- [ ] Do all new feature have an accompanying new test?
- [ ] Has the documentation been updated as necessary?

### CODE REVIEW

File: bindings/python/scripts/dlite-validate [hunk 1]
```diff
@@ -18,21 +18,34 @@ def parse(url, driver=None, options="mode=r", id=None):
         driver: The name of storage plugin to use.
         options: Options passed to the storage plugin.
         id: ID of instance to load if the storage contains multiple instances.
+            The default is to load all instances.
 
     Returns:
-        A new instance.
+        List of new instance.
```
Comment (jesper-friis) [inline bindings/python/scripts/dlite-validate:24]:
```suggestion
        List of new instances.
```

File: bindings/python/dlite-jstore-python.i [hunk 2]
```diff
@@ -0,0 +1,262 @@
...
+            is true if `single=True` and URI is None, otherwise it
+            is false.
+        with_meta: Whether to always include "meta" (even for metadata)
+        with_parent: Whether to include parent info for transactions.
+        urikey: Whether the URI is the preferred keys in multi-instance
+            format.
+
+    Notes:
+        This method works with the dict-representation and does not
+        access instances.  The only exception is when `d` correspond to
```
Comment (francescalb) [inline bindings/python/dlite-jstore-python.i:34]:
```suggestion
        access instances.  The only exception is when `d` corresponds to
```

File: bindings/python/tests/test_jstore.py [hunk 3]
```diff
@@ -0,0 +1,206 @@
...
+
+d1 = {
+    "uuid": "d6a1c1db-44b6-5b87-b815-83f1127395b6",
+    "meta": "http://onto-ns.com/meta/ex/0.2/Test",
+    "dimensions": {"n": 3},
+    "properties": {
+        "a": "hello",
+        "b": [1.1, 2.2, 3.3],
+    },
+}
```
Comment (francescalb) [inline bindings/python/tests/test_jstore.py:46]: What is the difference between these two?

File: bindings/python/tests/test_jstore.py [hunk 4]
```diff
@@ -0,0 +1,206 @@
...
+d2 = {
+    "uuid": "d6a1c1db-44b6-5b87-b815-83f1127395b6",
+    "meta": "http://onto-ns.com/meta/ex/0.2/Test",
+    "dimensions": {"n": 3},
+    "properties": {
+        "a": "hello",
+        "b": [1.1, 2.2, 3.3],
+    },
+}
```
Comment (francescalb) [inline bindings/python/tests/test_jstore.py:55]: What is the difference between these two?

Comment (jesper-friis) [reply bindings/python/tests/test_jstore.py]: Good observation. You are right, for data instances there are no difference between the soft5 and soft7 representation. I added a comment about this to the test

File: bindings/python/dlite-jstore-python.i [hunk 5]
```diff
@@ -0,0 +1,262 @@
...
+%pythoncode %{
+import json
+from typing import Mapping, Sequence
+
+
+def format_dict(
+    d, id=None, soft7=True, single=None, with_uuid=None, with_meta=False,
+    with_parent=True, urikey=False,
+):
+    """Return a copy of `d` formatted according to the given options.
```
Comment (francescalb) [inline bindings/python/dlite-jstore-python.i:14]: d should be a datamodel in dict-representation, right? We can't just give it any dict?

File: bindings/python/dlite-jstore-python.i [hunk 6]
```diff
@@ -0,0 +1,262 @@
...
+        with_meta: Whether to always include "meta" (even for metadata)
+        with_parent: Whether to include parent info for transactions.
+        urikey: Whether the URI is the preferred keys in multi-instance
+            format.
+
+    Notes:
+        This method works with the dict-representation and does not
+        access instances.  The only exception is when `d` corresponds to
+        a data instance who's dimensions is a list of dimension lengths.
+        In this case is the metadata needed to get dimension names.
```
Comment (francescalb) [inline bindings/python/dlite-jstore-python.i:36]:
```suggestion
        In this case the metadata is needed to get dimension names.
```

File: bindings/python/dlite-jstore-python.i (hunk 5 above)
Comment (jesper-friis) [reply bindings/python/dlite-jstore-python.i:14]: Yes, you are right. It should be a dict-representation of a datamodel.

**Review by francescalb — APPROVED** [review]

**Pull request merged**